pkg/image/controller/controller.go

```diff
@@ -39,35 +33,105 @@ const retryCount = 2
 // are left as is (those are updated through status).
+// There are 3 use cases here:
+// 1. spec.DockerImageRepository defined without any tags results in all tags being imported
+//    from default registry
+// 2. spec.DockerImageRepository + tags defined - import only the provided tags,
+//    for kind == DockerImage import underlying image, iow. exact tag (no provided means latest),
+//    for kind != DockerImage import tag from the default registry
+// 3. spec.DockerImageRepository not defined - import tags per its definition.
 func (c *ImportController) Next(stream *api.ImageStream) error {
 	if !needsImport(stream) {
 		return nil
 	}
+
 	name := stream.Spec.DockerImageRepository
+	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
+
+	var tags map[string]string
+	// read default image metadata
+	if len(name) != 0 {
+		var (
+			err   error
+			retry bool
+		)
+		tags, err, retry = c.getDefaultTags(name, insecure)
+		if err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	}
+	if len(stream.Spec.Tags) == 0 {
+		// no tags specified, import all default
+		if err, retry := c.importDefaultTags(stream, tags, insecure); err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	} else {
+		// import only specified tags
+		limitedTags := make(map[string]string)
+		for tagName, specTag := range stream.Spec.Tags {
+			if tag, ok := tags[tagName]; ok && specTag.From == nil {
+				limitedTags[tagName] = tag
+			}
+		}
+
+		if err, retry := c.importDefaultTags(stream, limitedTags, insecure); err != nil {
+			util.HandleError(err)
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+		if err, retry := c.importExternalTags(stream, insecure); err != nil {
+			if !retry {
+				return err
+			}
+			return c.done(stream, err.Error(), retryCount)
+		}
+	}
+
+	// we've completed our updates
+	return c.done(stream, "", retryCount)
+}
+
+// getDefaultTags reads all tags from the stream's default registry,
+// returns tags, an error and bool flag specifying if the error is retryable
+func (c *ImportController) getDefaultTags(name string, insecure bool) (map[string]string, error, bool) {
+	var tags map[string]string
 
 	ref, err := api.ParseDockerImageReference(name)
 	if err != nil {
 		err = fmt.Errorf("invalid docker image repository, cannot import data: %v", err)
-		util.HandleError(err)
-		return c.done(stream, err.Error(), retryCount)
+		return tags, err, true
 	}
 
-	insecure := stream.Annotations != nil && stream.Annotations[api.InsecureRepositoryAnnotation] == "true"
-
 	client := c.client
 	if client == nil {
 		client = dockerregistry.NewClient()
 	}
 	conn, err := client.Connect(ref.Registry, insecure)
 	if err != nil {
-		return err
+		return tags, err, false
```

Comment:
I didn't intend to change the current logic, so that's a question to @ncdc rather. On the other hand I can agree that retrying makes sense, unless there's something I don't know about connecting to image registry.